src/verifier/verifier.rs

```diff
@@ -19,7 +27,535 @@ pub fn build_symbol_table(ast: &mut Whamm, err: &mut ErrorGen) -> SymbolTable {
+                                    .map(|l| l.line_col),
+                            )
+                        }
+                    } else {
+                        self.err.type_check_error(
+                            false,
+                            format! {"Can't look up {} in symbol table", name},
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                    }
+                }
+
+                None
+            }
+            Expr::UnOp { op, expr, loc } => {
+                let expr_ty_op = self.visit_expr(expr);
+                if let Some(expr_ty) = expr_ty_op {
+                    match op {
+                        UnOp::Not => {
+                            if expr_ty == DataType::Boolean {
+                                Some(DataType::Boolean)
+                            } else {
+                                self.err.type_check_error(
+                                    false,
+                                    "Not operator can only be applied to boolean".to_owned(),
+                                    &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                        .map(|l| l.line_col),
+                                );
+                                None
+                            }
+                        }
+                    }
+                } else {
+                    self.err.type_check_error(
+                        false,
+                        "Can't get type of expr of this unary operation".to_owned(),
+                        &<Option<crate::parser::types::Location> as Clone>::clone(expr.loc())
+                            .map(|l| l.line_col),
+                    );
+                    None
+                }
+            }
+            Expr::Call {
+                fn_target,
+                args,
+                loc,
+            } => {
+                // TODO: finish type checking for function calls
+                // lookup type of function
+                let mut param_tys = vec![];
+                if let Some(args) = args {
+                    for arg in args {
+                        if let Some(ty) = self.visit_expr(arg) {
+                            param_tys.push(ty);
+                        } else {
+                            self.err.type_check_error(
+                                false,
+                                "Can't get type of argument".to_owned(),
+                                &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                    .map(|l| l.line_col),
+                            );
+                            return None;
+                        }
+                    }
+                } // else function has no arguments
+
+                let fn_name = match *fn_target.clone() {
```

Comment:
Why the clone?

Reply:
I was going to use `**fn_target` since I need to deref and unbox `fn_target` But it warns me:
```
error[E0507]: cannot move out of `fn_target.name` as enum variant `VarId` which is behind a shared reference
   --> src/verifier/verifier.rs:389:37
    |
389 |                 let fn_name = match **fn_target {
    |                                     ^^^^^^^^^^^
390 |                     Expr::VarId { name, .. } => name,
    |                                   ----
    |                                   |
    |                                   data moved here
    |                                   move occurs because `name` has type `String`, which does not implement the `Copy` trait
    |
help: consider removing the dereference here
    |
389 -                 let fn_name = match **fn_target {
389 +                 let fn_name = match *fn_target {
    |
```